core/src/apps/management/recovery_device/recover.py

```diff
@@ -1,80 +1,90 @@
 from trezor.crypto import bip39, slip39
-from trezor.errors import GroupThresholdReachedError, MnemonicError
+from trezor.errors import MnemonicError
 
 from apps.common import storage
 
 if False:
-    from typing import Optional
+    from typing import Optional, Tuple
 
 
 class RecoveryAborted(Exception):
     pass
 
 
-_GROUP_STORAGE_OFFSET = 16
-
-
 def process_bip39(words: str) -> bytes:
     """
     Receives single mnemonic and processes it. Returns what is then stored
     in the storage, which is the mnemonic itself for BIP-39.
     """
     if not bip39.check(words):
-        raise MnemonicError()
+        raise MnemonicError
     return words.encode()
 
 
-def process_slip39(words: str) -> Optional[bytes, int, int]:
+def process_slip39(words: str) -> Tuple[Optional[bytes], slip39.Share]:
     """
-    Receives single mnemonic and processes it. Returns what is then stored in storage or
-    None if more shares are needed.
+    Processes a single mnemonic share. Returns the encrypted master secret
+    (or None if more shares are needed) and the share's group index and member index.
     """
-    identifier, iteration_exponent, group_index, group_threshold, group_count, index, threshold, value = slip39.decode_mnemonic(
-        words
-    )  # TODO: use better data structure for this
+    share = slip39.decode_mnemonic(words)
 
     remaining = storage.recovery.fetch_slip39_remaining_shares()
-    index_with_group_offset = index + group_index * _GROUP_STORAGE_OFFSET
+    # TODO: move this whole logic to storage
+    index_with_group_offset = share.index + share.group_index * slip39.MAX_SHARE_COUNT
```

Comment:
We should probably do this before the freeze. But I can do it in the PR that will follow this one so we can get this beast merged finally. Other than this LGTM.